File: src/bridge/rw_mol.rs
#[cxx::bridge(namespace = "RDKit")]
pub mod ffi {
    unsafe extern "C++" {
        include!("wrapper/include/ro_mol.h");
        include!("wrapper/include/rw_mol.h");

        pub type RWMol;
        pub type ROMol = crate::ro_mol_ffi::ROMol;

        pub fn rw_mol_from_mol_block(
            mol_block: &CxxString,
            sanitize: bool,
            remove_hs: bool,
            strict_parsing: bool,
        ) -> SharedPtr<RWMol>;

        pub fn rw_mol_from_ro_mol(
            mol: &SharedPtr<ROMol>,
            quick_copy: bool,
            conf_id: i32,
        ) -> SharedPtr<RWMol>;

        pub fn rw_mol_from_rw_mol(mol: &SharedPtr<RWMol>) -> SharedPtr<RWMol>;

        pub fn rw_mol_to_ro_mol(mol: SharedPtr<RWMol>) -> SharedPtr<ROMol>;

        pub fn smarts_to_mol(smarts: &CxxString) -> Result<SharedPtr<RWMol>>;
    }
}

File: wrapper/src/rw_mol.cc
#include "rust/cxx.h"
#include <GraphMol/GraphMol.h>
#include <RDGeneral/FileParseException.h>
#include <GraphMol/FileParsers/FileParsers.h>
#include <GraphMol/SmilesParse/SmilesParse.h>

namespace RDKit {
    std::shared_ptr<RWMol> rw_mol_from_mol_block(const std::string &mol_block, bool sanitize, bool remove_hs, bool strict_parsing) {
        RWMol *mol;
        try {
            mol = MolBlockToMol(mol_block, sanitize, remove_hs, strict_parsing);
        } catch (const RDKit::FileParseException &e) {
            mol = nullptr;
        } catch (const RDKit::AtomValenceException &e) {
            mol = nullptr;
        }
        return std::shared_ptr<RWMol>(mol);
    }

    std::shared_ptr<RWMol> rw_mol_from_ro_mol(const std::shared_ptr<ROMol> &mol, bool quick_copy, int conf_id) {
        RWMol *rw_mol = new RWMol(*mol, quick_copy, conf_id);
        return std::shared_ptr<RWMol>(rw_mol);
    }

    std::shared_ptr<RWMol> rw_mol_from_rw_mol(const std::shared_ptr<RWMol> &mol) {
        RWMol *rw_mol = new RWMol(*mol);
        return std::shared_ptr<RWMol>(rw_mol);
    }

    std::shared_ptr<ROMol> rw_mol_to_ro_mol(std::shared_ptr<RWMol> mol) {
        return std::static_pointer_cast<ROMol>(mol);
    }

    std::shared_ptr<RWMol> smarts_to_mol(const std::string &smarts) {
        return std::shared_ptr<RWMol>(SmartsToMol(smarts));
    }
}